cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/repo_auth.go

```diff
@@ -318,96 +319,149 @@ func (s *Server) getRepoTlsConfigAndAuth(ctx context.Context, repo sourcev1.Helm
-				secret = newLocalOpaqueSecret(repoName)
+				return nil, false, status.Errorf(codes.InvalidArgument, "Invalid configuration, unexpected REDACTED content")
 			}
+		} else {
+			secret.Data["caFile"] = []byte(caCert)
+			isSameSecret = false
+		}
+	} else {
+		if hadSecretTlsCa {
+			isSameSecret = false
 		}
+	}
+
+	if auth != nil && auth.Type != corev1.PackageRepositoryAuth_PACKAGE_REPOSITORY_AUTH_TYPE_UNSPECIFIED {
 		switch auth.Type {
 		case corev1.PackageRepositoryAuth_PACKAGE_REPOSITORY_AUTH_TYPE_BASIC_AUTH:
-			if unp := auth.GetUsernamePassword(); unp != nil {
-				if unp.Username == redactedString && unp.Password == redactedString {
-					isSameSecret = true
-				} else {
+			unp := auth.GetUsernamePassword()
+			if unp == nil || unp.Username == "" || unp.Password == "" {
+				return nil, false, status.Errorf(codes.InvalidArgument, "Username/Password configuration is missing")
+			}
+			if (unp.Username == redactedString && !hadSecretUsername) || (unp.Password == redactedString && !hadSecretPassword) {
+				return nil, false, status.Errorf(codes.InvalidArgument, "Invalid configuration, unexpected REDACTED content")
+			}
+
+			if existingSecret != nil {
+				secret.Data["username"] = existingSecret.Data["username"]
+				secret.Data["password"] = existingSecret.Data["password"]
+			}
+			if unp.Username != redactedString || unp.Password != redactedString {
+				isSameSecret = false
+				if unp.Username != redactedString {
 					secret.Data["username"] = []byte(unp.Username)
+				}
+				if unp.Password != redactedString {
 					secret.Data["password"] = []byte(unp.Password)
 				}
-			} else {
-				return nil, false, status.Errorf(codes.Internal, "Username/Password configuration is missing")
 			}
 		case corev1.PackageRepositoryAuth_PACKAGE_REPOSITORY_AUTH_TYPE_TLS:
 			if repoType == sourcev1.HelmRepositoryTypeOCI {
 				// ref https://fluxcd.io/flux/components/source/helmrepositories/#tls-authentication
 				// Note: TLS authentication is not yet supported by OCI Helm repositories.
 				return nil, false, status.Errorf(codes.Internal, "Package repository authentication type %q is not supported for OCI repositories", auth.Type)
-			} else {
-				if ck := auth.GetTlsCertKey(); ck != nil {
-					if ck.Cert == redactedString && ck.Key == redactedString {
-						isSameSecret = true
-					} else {
-						secret.Data["certFile"] = []byte(ck.Cert)
-						secret.Data["keyFile"] = []byte(ck.Key)
-					}
-				} else {
-					return nil, false, status.Errorf(codes.Internal, "TLS Cert/Key configuration is missing")
-				}
 			}
 
+			ck := auth.GetTlsCertKey()
+			if ck == nil || ck.Cert == "" || ck.Key == "" {
+				return nil, false, status.Errorf(codes.InvalidArgument, "TLS Cert/Key configuration is missing")
+			}
+			if (ck.Cert == redactedString && !hadSecretTlsCert) || (ck.Key == redactedString && !hadSecretTlsKey) {
+				return nil, false, status.Errorf(codes.InvalidArgument, "Invalid configuration, unexpected REDACTED content")
+			}
```

Comment:
Thanks for all these checks - ensuring that we'll know if the UX sends the incorrect redacted content (I assume?).